velox/functions/sparksql/fuzzer/SparkQueryRunner.cpp

```diff
@@ -0,0 +1,388 @@
+    // Spark aggregation does not support sort keys.
+    if (!aggregate.sortingKeys.empty()) {
+      return false;
+    }
+    const auto functionName = aggregate.call->name();
+    if (functionName == "max_by" || functionName == "min_by") {
+      VELOX_CHECK_EQ(aggregate.rawInputTypes.size(), 2);
+      if (aggregate.rawInputTypes[1]->kind() == TypeKind::MAP) {
+        // Spark's `max_by` and `min_by` do not support ordering on map type.
+        return false;
+      }
+    }
+    if (aggregate.distinct) {
+      for (const auto& inputType : aggregate.rawInputTypes) {
+        // In Spark, grouping keys cannot be map type.
+        if (inputType->kind() == TypeKind::MAP) {
+          return false;
+        }
+      }
+    }
+  }
+  for (const auto& key : aggregationNode->groupingKeys()) {
+    // In Spark, grouping keys cannot be map type.
+    if (key->type()->kind() == TypeKind::MAP) {
+      return false;
+    }
+  }
+  return true;
+}
+
+std::optional<std::string> SparkQueryRunner::toSql(
+    const std::shared_ptr<const core::AggregationNode>& aggregationNode) {
+  if (!supported(aggregationNode)) {
+    return std::nullopt;
+  }
+
+  // Assume plan is Aggregation over Values.
+  VELOX_CHECK(aggregationNode->step() == core::AggregationNode::Step::kSingle);
+
+  std::vector<std::string> groupingKeys;
+  for (const auto& key : aggregationNode->groupingKeys()) {
+    groupingKeys.push_back(key->name());
+  }
+
+  std::stringstream sql;
+  sql << "SELECT " << folly::join(", ", groupingKeys);
+
+  const auto& aggregates = aggregationNode->aggregates();
+  if (!aggregates.empty()) {
+    if (!groupingKeys.empty()) {
+      sql << ", ";
+    }
+
+    for (auto i = 0; i < aggregates.size(); ++i) {
+      appendComma(i, sql);
+      const auto& aggregate = aggregates[i];
+      sql << toAggregateCallSql(aggregate.call, aggregate.distinct);
+
+      if (aggregate.mask != nullptr) {
+        sql << " filter (where " << aggregate.mask->name() << ")";
+      }
+      sql << " as " << aggregationNode->aggregateNames()[i];
+    }
+  }
+
+  sql << " FROM tmp";
+
+  if (!groupingKeys.empty()) {
+    sql << " GROUP BY " << folly::join(", ", groupingKeys);
+  }
+
+  return sql.str();
+}
+
+std::optional<std::string> SparkQueryRunner::toSql(
+    const std::shared_ptr<const core::ProjectNode>& projectNode) {
+  auto sourceSql = toSql(projectNode->sources()[0]);
+  if (!sourceSql.has_value()) {
+    return std::nullopt;
+  }
+
+  std::stringstream sql;
+  sql << "SELECT ";
+
+  for (auto i = 0; i < projectNode->names().size(); ++i) {
+    appendComma(i, sql);
+    auto projection = projectNode->projections()[i];
+    if (auto field =
+            std::dynamic_pointer_cast<const core::FieldAccessTypedExpr>(
+                projection)) {
+      sql << field->name();
+    } else if (
+        auto call =
+            std::dynamic_pointer_cast<const core::CallTypedExpr>(projection)) {
+      sql << toCallSql(call);
+    } else {
+      VELOX_NYI();
```

Comment:
nit: perhaps, add an error message and include the 'projectNode' info to help troubleshooting,